TFNetworkRecLayer.py

```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+      size_placeholder = sources[1].output.size_placeholder.copy()
+
+    beam_size = sources[0].output.beam_size
+    name = kwargs["name"]
+    dtype = "float32"
+    available_for_inference = all([(src.output.available_for_inference) for src in sources])
+
+    return Data(
+      name="%s_output" % name,
+      shape=shape,
+      batch_dim_axis=batch_dim_axis,
+      time_dim_axis=time_dim_axis,
+      size_placeholder=size_placeholder,
+      available_for_inference=available_for_inference,
+      dtype=dtype,
+      beam_size = beam_size,
+      sparse=False)
+
+  def _add_params(self, scope_name_prefix, params):
+    """
+    :param str scope_name_prefix:
+    :param list[tf.Variable] params:
+    """
+    for p in params:
+      if not p.name.startswith(scope_name_prefix):
+        continue
+      assert p.name.startswith(scope_name_prefix) and p.name.endswith(":0")
+      self.params[p.name[len(scope_name_prefix):-2]] = p
+
+  def _get_input(self):
+    """
+    :return: (x, seq_len), where x is (time,batch,...,dim) and seq_len is (batch,)
+    :rtype: (tf.Tensor, tf.Tensor)
+    """
+    assert len(self.sources) == 2
+    assert self.sources[0].output
+    assert self.sources[1].output
+    x = self.sources[0].output.get_placeholder_as_time_major()  # (time,batch,[dim])
+    seq_len_src = self.sources[0].output.get_sequence_lengths()
+
+    return x, seq_len_src
+
+  def get_constraints_value(self):
+    v = super(TwoDLSTMLayer, self).get_constraints_value()
+    from TFUtil import optional_add
+    if isinstance(self.cell, _SubnetworkRecCell):
+      for layer in self.cell.net.layers.values():
+        v = optional_add(v, layer.get_constraints_value())
+    return v
+
+  def _get_cell(self, unit_opts=None):
+    """
+    :param None|dict[str] unit_opts:
+    :rtype: _SubnetworkRecCell|tensorflow.contrib.rnn.RNNCell|tensorflow.contrib.rnn.FusedRNNCell|TFNativeOp.RecSeqCellOp
+    """
+    import TFNativeOp
+    rnn_cell_class = TFNativeOp.TwoDNativeLstmCell
+    n_hidden = self.output.dim
+    if unit_opts is None:
+      unit_opts = {}
+
+    assert not self.sources[0].output.sparse
+    n_input_dim_parts = [self.sources[0].output.dim, self.sources[1].output.dim]
+    cell = rnn_cell_class(
+      n_hidden=n_hidden, n_input_dim=sum(n_input_dim_parts), n_input_dim_parts=n_input_dim_parts,
+      input_is_sparse=self.sources[0].output.sparse,
+      pooling=self.pooling,
+      **unit_opts)
+    return cell
+
+  @classmethod
+  def get_rec_initial_extra_outputs(cls, batch_dim, rec_layer, **kwargs):
+    # state and output are dummy values that aren't used
+    # we have to specifiy them (and they must have 4 dims), because during inference (when multiple iterations
+    # for one 2DLSTM happen, as opposed to the training where everything is computed at once) we pass the intermediate
+    # values from one iteration to the other. 42 is added to show from the value of the variable that it's not
+    # supposed to be used.
```

Comment:
The values are used. They are the initial values.
In case they are not needed (training), do not set them.
Again, see other layers as example, e.g. `SelfAttentionLayer`.